packages/cli/src/serve.test.ts

```diff
@@ -0,0 +1,160 @@
+    killRegex?: RegExp;
+    signalRegex?: RegExp[];
+  },
+): Promise<{
+  exit: number;
+  stdout: string;
+  stderr: string;
+}> => {
+  return new Promise((resolve, reject) => {
+    const stdout: Buffer[] = [];
+    const stderr: Buffer[] = [];
+
+    const shell = process.platform === 'win32';
+    const proc = spawn(command, args, { ...options, shell });
+
+    proc.stdout?.on('data', data => {
+      stdout.push(Buffer.from(data));
+    });
+
+    proc.stderr?.on('data', data => {
+      stderr.push(Buffer.from(data));
+    });
+
+    /**
+     * Set an interval to check if we should kill the process.
+     * This was the easiest way I could think of of testing across two processes.
+     */
+    let intervalId: NodeJS.Timer | undefined = undefined;
+    if (eventConfig) {
+      intervalId = setInterval(() => {
+        const stdoutStr = Buffer.concat(stdout).toString('utf8');
+        if (eventConfig.killRegex?.test(stdoutStr)) {
+          proc.kill('SIGINT');
+        } else if (
+          eventConfig.signalRegex?.some(regex => regex.test(stdoutStr))
+        ) {
+          events?.emit('hit');
+        }
+      }, 1000);
+    }
+
+    const clearEventInterval = () => {
+      if (intervalId) {
+        try {
+          clearInterval(intervalId);
+        } catch (err) {
+          console.error(err);
+        }
+      }
+    };
+
+    /**
+     * Need a way to kill the process from another process.
+     */
+    events?.on('stop', signal => {
+      clearEventInterval();
+      proc.kill(signal);
+    });
+
+    proc.on('error', (...errorArgs) => {
+      clearEventInterval();
+      reject(errorArgs);
+    });
+    proc.on('exit', code => {
+      clearEventInterval();
+      resolve({
+        exit: code ?? 0,
+        stdout: Buffer.concat(stdout).toString('utf8'),
+        stderr: Buffer.concat(stderr).toString('utf8'),
+      });
+    });
+  });
+};
+
+const testProjectDir = path.resolve(
+  __dirname,
+  '__fixtures__/test-project/packages/app',
+);
+
+describe('end-to-end', () => {
+  const entryPoint = path.resolve(__dirname, '../bin/backstage-cli');
+
+  it('builds frontend with correct url overrides', async () => {
+    const buildProc = await executeCommand(
+      entryPoint,
+      ['package', 'build', '--public-path', '/test', '--backend-url', '/api'],
+      {
+        cwd: testProjectDir,
+      },
+    );
+    expect(buildProc.stderr).toContain(
+      'Loaded config from app-config.yaml, cli',
+    );
+
+    expect(buildProc.exit).toEqual(0);
+  });
+
+  it('starts frontend on correct url', async () => {
+    const startEmitter = new EventEmitter();
+    const frontendPort = await getPort();
+    startEmitter.on('hit', async () => {
+      const response = await fetch(
+        `http://localhost:${frontendPort}/test/catalog`,
+      );
+      const text = await response.text();
+      startEmitter.emit('stop', 'SIGINT');
+      expect(response.status).toBe(200);
+      expect(text.length).toBeGreaterThan(0);
```

Comment:
I think we could do a bit of extra validation here, for example just have something in `index.html` that we search for.

We also can't be sure that we actually reach this point in the test, it's best to have an `expect.assertions(...)` at the top of the test.